File: apps/android/src/test/java/com/tilawah/android/features/library/LibraryViewModelTests.kt
package com.tilawah.android.features.library

import app.cash.turbine.test
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import com.tilawah.android.app.AppError
import com.tilawah.android.backend.MockBackend
import com.tilawah.android.backend.SurahSummary
import com.tilawah.android.telemetry.TelemetrySpy

class LibraryViewModelTests {

    @Test
    fun `load transitions Idle to Loaded on success and emits library_fetch_succeeded`() = runTest {
        val backend = MockBackend(surahsResult = Result.success(fixtures))
        val telemetry = TelemetrySpy()
        val viewModel = LibraryViewModel(backend, telemetry)

        viewModel.state.test {
            assertEquals(LibraryUiState.Idle, awaitItem())
            viewModel.loadInternal()
            assertEquals(LibraryUiState.Loading, awaitItem())
            val loaded = awaitItem() as LibraryUiState.Loaded
            assertEquals(listOf("1", "2"), loaded.surahs.map { it.id })
            cancelAndIgnoreRemainingEvents()
        }
        assertTrue(telemetry.eventNames.contains("library.fetch.succeeded"))
    }

    @Test
    fun `load transitions to Failed on AppError network and emits library_fetch_failed`() = runTest {
        val backend = MockBackend(
            surahsResult = Result.failure(AppError.Network(IllegalStateException("offline"))),
        )
        val telemetry = TelemetrySpy()
        val viewModel = LibraryViewModel(backend, telemetry)

        viewModel.loadInternal()
        val failed = viewModel.state.value as LibraryUiState.Failed
        assertEquals("network", failed.error.telemetryCode)
        assertTrue(telemetry.eventNames.contains("library.fetch.failed"))
        val errorRecord = telemetry.records.filterIsInstance<TelemetrySpy.Record.Error>().single()
        assertEquals("network", errorRecord.code)
        assertEquals(mapOf("screen" to "library"), errorRecord.context)
    }

    @Test
    fun `surahOpened emits library_surah_opened with surah_id`() {
        val telemetry = TelemetrySpy()
        val viewModel = LibraryViewModel(MockBackend(), telemetry)
        viewModel.surahOpened(fixtures.first())
        val event = telemetry.records.filterIsInstance<TelemetrySpy.Record.Event>().single()
        assertEquals("library.surah.opened", event.name)
        assertEquals("1", event.attributes["surah_id"])
    }

    private val fixtures = listOf(
        SurahSummary("1", "الفاتحة", "Al-Fatihah", "Mecca", 7),
        SurahSummary("2", "البقرة", "Al-Baqarah", "Medina", 286),
    )
}

File: apps/ios/Tests/QuranRecitationAppTests/LibraryViewModelTests.swift
import XCTest
@testable import QuranRecitationApp

@MainActor
final class LibraryViewModelTests: XCTestCase {
  func test_load_publishesLoadedStateOnSuccess() async {
    let backend = MockBackend(surahsResult: .success(Self.fixtures))
    let telemetry = TelemetrySpy()
    let viewModel = LibraryViewModel(backend: backend, telemetry: telemetry)

    await viewModel.load()

    if case let .loaded(items) = viewModel.state {
      XCTAssertEqual(items.map(\.id), ["1", "2"])
    } else {
      XCTFail("expected .loaded, got \(viewModel.state)")
    }
    XCTAssertTrue(telemetry.eventNames().contains("library.fetch.succeeded"))
  }

  func test_load_publishesFailedStateOnNetworkError() async {
    let backend = MockBackend(surahsResult: .failure(.network(underlying: NSError(domain: "x", code: 0))))
    let telemetry = TelemetrySpy()
    let viewModel = LibraryViewModel(backend: backend, telemetry: telemetry)

    await viewModel.load()

    if case .failed(let error) = viewModel.state {
      XCTAssertEqual(error.telemetryCode, "network")
    } else {
      XCTFail("expected .failed, got \(viewModel.state)")
    }
    XCTAssertTrue(telemetry.eventNames().contains("library.fetch.failed"))
  }

  func test_surahOpened_emitsTelemetry() {
    let telemetry = TelemetrySpy()
    let viewModel = LibraryViewModel(backend: MockBackend(), telemetry: telemetry)

    viewModel.surahOpened(Self.fixtures[0])

    let openEvents = telemetry.records.filter {
      if case let .event(name, _) = $0 { return name == "library.surah.opened" }
      return false
    }
    XCTAssertEqual(openEvents.count, 1)
  }

  func test_filteredSurahs_appliesQueryAndChip() async {
    let backend = MockBackend(surahsResult: .success(Self.filteringFixtures))
    let viewModel = LibraryViewModel(backend: backend, telemetry: NoOpTelemetry())
    await viewModel.load()

    XCTAssertEqual(viewModel.filteredSurahs.count, 4, "no filters → all surahs")

    viewModel.filter = .mecca
    XCTAssertEqual(viewModel.filteredSurahs.map(\.id), ["1", "112"], "Meccan filter")

    viewModel.filter = .medina
    XCTAssertEqual(viewModel.filteredSurahs.map(\.id), ["2", "3"], "Medinan filter")

    viewModel.filter = .short
    XCTAssertEqual(viewModel.filteredSurahs.map(\.id), ["1", "112"], "≤20 ayahs")

    viewModel.filter = .all
    viewModel.query = "fati"
    XCTAssertEqual(viewModel.filteredSurahs.map(\.id), ["1"], "case-insensitive English match")

    viewModel.query = "ال"
    XCTAssertGreaterThan(viewModel.filteredSurahs.count, 0, "Arabic substring match")

    viewModel.query = "no-such-surah"
    XCTAssertTrue(viewModel.filteredSurahs.isEmpty, "no match → empty")
  }

  private static let filteringFixtures: [SurahSummary] = [
    SurahSummary(id: "1", nameAr: "الفاتحة", nameEn: "Al-Fatihah", ayahCount: 7, revelationPlace: "Mecca"),
    SurahSummary(id: "2", nameAr: "البقرة", nameEn: "Al-Baqarah", ayahCount: 286, revelationPlace: "Medina"),
    SurahSummary(id: "112", nameAr: "الإخلاص", nameEn: "Al-Ikhlas", ayahCount: 4, revelationPlace: "Mecca"),
    SurahSummary(id: "3", nameAr: "آل عمران", nameEn: "Aal-Imran", ayahCount: 200, revelationPlace: "Medina")
  ]

  // MARK: - Fixtures

  private static let fixtures: [SurahSummary] = [
    SurahSummary(id: "1", nameAr: "الفاتحة", nameEn: "Al-Fatihah", ayahCount: 7, revelationPlace: "Mecca"),
    SurahSummary(id: "2", nameAr: "البقرة", nameEn: "Al-Baqarah", ayahCount: 286, revelationPlace: "Medina")
  ]
}
